app/javascript/src/Login/components/LoginForm.tsx

```diff
@@ -0,0 +1,127 @@
+import { useState } from 'react'
+import {
+  ActionGroup,
+  Button,
+  Form,
+  FormGroup,
+  HelperText,
+  HelperTextItem,
+  TextInput
+} from '@patternfly/react-core'
+
+import { validateLogin } from 'Login/utils/validations'
+import { CSRFToken } from 'utilities/CSRFToken'
+
+import type { FlashMessage } from 'Types/FlashMessages'
+import type { FunctionComponent } from 'react'
+
+interface Props {
+  error?: FlashMessage;
+  providerSessionsPath: string;
+  session: {
+    username: string | null;
+  };
+}
+
+const LoginForm: FunctionComponent<Props> = ({
+  error,
+  providerSessionsPath,
+  session
+}) => {
+  const [username, setUsername] = useState(session.username ?? '')
+  const [password, setPassword] = useState('')
+  const [validationVisibility, setValidationVisibility] = useState({
+    username: false,
+    password: false
+  })
+
+  const onUsernameChange = (value: string) => {
+    setUsername(value)
+    setValidationVisibility(prev => ({ ...prev, username: false }))
+  }
+
+  const onPasswordChange = (value: string) => {
+    setPassword(value)
+    setValidationVisibility(prev => ({ ...prev, password: false }))
+  }
```

Comment:
Somehow move this into custom hook... create an object of `onChange`s and another of `onBlur`s?